### PR TITLE
Tb main vcf qc rename samples

### PR DESCRIPTION
**Pull request by TedBrookings**
Add utility to optionally rename input VCF sample IDs to conform to comparison data set.

### CODE REVIEW

**Comment by TedBrookings** [comment]
I added an input file `MainVcfQc_gq_filter.json.tmpl`, which I used to run QC on my input data. I've left it in for now so that a) reviewers can see how the new arguments are intended to be filled out, and b) I can re-run the job if there are any problems. I think I should probably remove it before merging. I'm not sure if the altered values should be propagated into the main `MainVcfQc.json.tmpl`

File: inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl
Comment (epiercehoffman) [inline inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl:23]: I think it's cleaner to have one JSON template in general, but if this one has trios then maybe we should keep it as an alternative template to test the full breadth of functionality of MainVcfQc. Is this data able to be shared publicly? If so, can you make this VCF & PED file public? And can you move the VCF to a bucket without a TTL?

Also, can you rename the file to something like `MainVcfQc.WithTrios.json.tmpl` to be descriptive of what it contains & follow the naming convention of `WDLName.Descriptor.json.tmpl` (like GATKSVPipelineBatch.FromSampleEvidence.json.tmpl)?

File: inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl
Comment (epiercehoffman) [inline inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl:10]: Can you make this reference the values JSON - similar to the site-level ones above?

File: wdl/Utils.wdl
Comment (epiercehoffman) [inline wdl/Utils.wdl:819]: It seems like it would be simpler to just take this TSV sample ID map as the input instead - what do you think? Personally I prefer the TSV because I think it's easier to generate for an analyst, and it removes some computation from this task

File: wdl/TasksMakeCohortVcf.wdl
Comment (epiercehoffman) [inline wdl/TasksMakeCohortVcf.wdl:999]: Definitely more readable with non-overlapping variable names this way

File: wdl/MainVcfQc.wdl
Comment (epiercehoffman) [inline wdl/MainVcfQc.wdl:91]: Do you think there's a way to do the renaming for just the external per-sample benchmarking, but not for the rest of the per-sample QC collection? That might be too complicated, because there's a lot more files, but if it was able to just map between the current cohort's sample IDs and the external cohort's sample IDs within the per-sample comparison script that seems like it would be ideal

File: inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl
Comment (epiercehoffman) [inline inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl:13]: Does this input also need to be added to MakeCohortVcf, the Batch WDL, and the single-sample WDL? Or maybe just use sv_pipeline_docker to avoid having to do that?

File: wdl/MainVcfQc.wdl
Comment (epiercehoffman) [inline wdl/MainVcfQc.wdl:19]: Should make this an option in all workflows that call MainVcfQc

**Review by epiercehoffman — CHANGES_REQUESTED** [review]
Good idea to add a sample renaming function to the sample-level comparisons here - this is a pain otherwise. 

Most of my comments are more like suggestions/ideas and may turn out to not be practical, but I wanted to bring them up at least. The change that I definitely want to request before you merge is to move the VCF in the new JSON template to a bucket without a TTL.

File: wdl/Utils.wdl
Comment (TedBrookings) [reply wdl/Utils.wdl:819]: bcftools is taking as input just a list of the new sample IDs, which it will use based on position alone (no mapping involved). The issue I faced in designing this is that we have potentially multiple input VCFs, in which case I could either have multiple input sample-renaming files, or one master input file that we could subset. I decided the first approach was actually harder to deal with, so I went with the later. Both have their downsides though. That said, I think the computation is only adding mental overhead for understanding this task, I don't think it adds meaningfully to runtime.

I actually developed a script to produce this input file, which I considered adding to the PR, but I didn't call it in a WDL. I could add it to the `scripts` folder though.

File: wdl/MainVcfQc.wdl
Comment (TedBrookings) [reply wdl/MainVcfQc.wdl:91]: Honestly I think the thing to do is to rename the samples in the external comparison set: I went with this approach because it was suggested to me and I was too busy to sit and think through the implications. The problem with only renaming during the external per-sample benchmarking, is that it takes its data from the bed files produced by `CollectQcVcfWide`. It felt like I might have to intervene in names in multiple places, and maybe mess up somewhere.

If this is a significant problem, it could be a good reason to go back to the drawing board and instead rename the files in the external tarball.

File: inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl
Comment (TedBrookings) [reply inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl:13]: We could use anything derived from `samtools_cloud_docker`, so `sv_pipeline_docker` would work. I went with `samtools_cloud_docker` because it's smaller, but I don't think that's a huge deal. I think the best way forward would be to keep `samtools_cloud_docker` as the input for the Task in `Utils.wdl`, but then just pass `sv_pipeline_docker` as the corresponding input from `MainVcfQc.wdl`. I'll make that change unless you object.

File: wdl/MainVcfQc.wdl
Comment (TedBrookings) [reply wdl/MainVcfQc.wdl:19]: Done. I omitted Module07MinGqStep2MergePCRStatus because a) I'm suspicious of it, and b) it doesn't have the option to run per-sample comparisons anyway.

File: inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl
Comment (epiercehoffman) [reply inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl:13]: Sounds like a good plan to me

File: wdl/Utils.wdl
Comment (epiercehoffman) [reply wdl/Utils.wdl:819]: bcftools also has the option to take a whitespace-separated file with a column for the old ID & a column for the new ID, so that's what I was referring to - sorry for the confusion since you were using a different format here. If we stick with the VCF reheader approach, my personal preference would be to provide a TSV in that format as input, rather than a JSON, because it's super easy to generate for an analyst working in excel or with Terra data tables. 

My understanding is that the multiple input VCFs should contain the same samples, but they may be sharded by genomic position, so I think it's ok to have one reheader file for all VCFs.

File: wdl/MainVcfQc.wdl
Comment (epiercehoffman) [reply wdl/MainVcfQc.wdl:91]: Do you think it could work to pass a sample renaming map to this script and use the ID that matches the external benchmarking dataset here when creating the per-sample BED files in the tarball? https://github.com/broadinstitute/gatk-sv/blob/cd5ce7e0a890c4c5234341e9fc976b4d56d3ca60/src/sv-pipeline/scripts/vcf_qc/collectQC.perSample_benchmarking.sh#L158

**Review by mwalker174 — CHANGES_REQUESTED** [review]
Thanks @TedBrookings I thought you could just do this as a one-off for the hgdp call set, but working it into the pipeline is really above and beyond. This will save us a lot of time in the future!

For general use, I think we should be careful since some call sets contain problematic external IDs, so we should stick to the internal IDs where we can. I also need to double-check that we can share the hgdp vcf.

File: inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl
Comment (TedBrookings) [reply inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl]: I renamed it to `MainVcfQc_per_sample_benchmark.json.tmpl`, because adding the per-sample benchmarking is the main difference. However, it's also using a different VCF and PED file than the other template too. Since I've had quite variable problems with memory and disk, I think it's possible the attribute overrides relate to those inputs.

File: inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl
Comment (TedBrookings) [reply inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl]: Done

File: wdl/Utils.wdl
Comment (TedBrookings) [reply wdl/Utils.wdl:819]: Used a totally different strategy (as discussed), but the input is now a TSV. The script that produces it is `gatk-sv/scripts/inputs/get_rename_benchmark_samples_map.py`

File: inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl
Comment (TedBrookings) [reply inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl:13]: I followed this plan. However it's now moot: the task for renaming VCF samples is not called. I'm inclined to delete it, since it is not used and doesn't use the preferred TSV input format. However, I figured I would wait for reviewer confirmation.

File: wdl/Utils.wdl
Comment (TedBrookings) [reply wdl/Utils.wdl:801]: This is probably moot because I expect you will want the task removed, but I want to discuss for future reference. I wrote this in mind with local execution / non-gcp stuff in mind. The token would be empty in that case, but since the files are local, their path would be good and I expect the rest of the task would execute. Is this just too risky a strategy anyway?

File: inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl
Comment (epiercehoffman) [reply inputs/templates/test/MainVcfQc/MainVcfQc_gq_filter.json.tmpl]: Sure, per_sample_benchmark is fine with me as an identifier. Can you still rename it to `MainVcfQc.per_sample_benchmark.json.tmpl` to match the `.` format used in other JSON templates in our repo?

File: wdl/MainVcfQc.wdl
Comment (TedBrookings) [reply wdl/MainVcfQc.wdl]: Actually I agree, and it wasn't a tough fix. I'm running a quick test to make sure that nothing crazy happens, but it shouldn't delay anything.

File: inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl
Comment (epiercehoffman) [reply inputs/templates/terra_workspaces/cohort_mode/workflow_configurations/MainVcfQc.json.tmpl:13]: Yes, please go ahead and delete this input as well as the old task

**Review by epiercehoffman — APPROVED** [review]
Thanks for making the requested changes! Approved pending removal of samtools_cloud_docker input from MainVcfQc Terra JSON template & old renaming task that is no longer used

File: wdl/Utils.wdl
Comment (mwalker174) [reply wdl/Utils.wdl]: That's a fair point that it would just ignore the token in local mode, so fine to leave in. I think we had the same discussion in another PR now that I think of it.

**Review by mwalker174 — APPROVED** [review]

**Pull request merged**
